src/controller/datacontroller.js

```diff
@@ -198,10 +204,28 @@ export default class DataController {
 			this.model.selection.removeAllRanges();
 			this.model.selection.clearAttributes();
 
+			// Parse data to model and extract markers from parsed document fragment.
+			const { documentFragment, markersData } = extractMarkersFromModelFragment( this.parse( data ) );
+
 			// Initial batch should be ignored by features like undo, etc.
-			this.model.batch( 'transparent' )
+			const batch = this.model.batch( 'transparent' );
+
+			// Replace current editor data by the new one.
+			batch
 				.remove( ModelRange.createIn( modelRoot ) )
-				.insert( ModelPosition.createAt( modelRoot, 0 ), this.parse( data ) );
+				.insert( ModelPosition.createAt( modelRoot, 0 ), documentFragment );
+
+			// Add markers to the document.
+			for ( const marker of markersData ) {
+				const markerName = marker[ 0 ];
+				const markerData = marker[ 1 ];
+				const range = new ModelRange(
+					new ModelPosition( modelRoot, markerData.startPath ),
+					markerData.endPath ? new ModelPosition( modelRoot, markerData.endPath ) : null
+				);
+
+				batch.setMarker( this.model.markers.set( markerName, range ) );
```

Comment:
AFAIR you can directly use `setMarker` - no need to add that marker to `MarkerCollection` first. `MarkerOperation`, upon execution, will do it.